mteb/models/openai_models.py

```diff
@@ -15,38 +15,91 @@
 
 
 class OpenAIWrapper(Wrapper):
-    def __init__(self, model_name: str, embed_dim: int | None = None, **kwargs) -> None:
+    def __init__(
+        self,
+        model_name: str,
+        max_tokens: int,
+        tokenizer_name: str = "cl100k_base",  # since all models use this tokenizer now
+        embed_dim: int | None = None,
+        **kwargs,
+    ) -> None:
+        """Wrapper for OpenAIs embedding API.
+        To handle documents larger than 8192 tokens, we truncate the document to the specified sequence length.
+        """
         requires_package(self, "openai", "Openai text embedding")
         from openai import OpenAI
 
+        requires_package(self, "tiktoken", "Tiktoken package")
+        import tiktoken
+
         self._client = OpenAI()
         self._model_name = model_name
         self._embed_dim = embed_dim
+        self._max_tokens = max_tokens
+        self._encoding = tiktoken.get_encoding(tokenizer_name)
+
+    def truncate_text_tokens(self, text):
+        """Truncate a string to have `max_tokens` according to the given encoding."""
+        truncated_sentence = self._encoding.encode(text)[: self._max_tokens]
+        return self._encoding.decode(truncated_sentence)
 
     def encode(self, sentences: list[str], **kwargs: Any) -> np.ndarray:
         requires_package(self, "openai", "Openai text embedding")
+
         from openai import NotGiven
 
         if self._model_name == "text-embedding-ada-002" and self._embed_dim is not None:
             logger.warning(
                 "Reducing embedding size available only for text-embedding-3-* models"
             )
 
+        trimmed_sentences = []
+        for sentence in sentences:
+            encoded_sentence = self._encoding.encode(sentence)
+            if len(encoded_sentence) > self._max_tokens:
+                truncated_sentence = self.truncate_text_tokens(sentence)
+                trimmed_sentences.append(truncated_sentence)
+            else:
+                trimmed_sentences.append(sentence)
+
         max_batch_size = 2048
```

Comment:
Should we change this to something like `max_batch_size = kwargs.get("batch_size", 2048)`?